src/emulator/types.ts

```diff
@@ -57,20 +76,70 @@ export interface JavaEmulatorCommand {
   binary: string;
   args: string[];
   optionalArgs: string[];
+  joinArgs: boolean;
 }
 
-export interface JavaEmulatorDetails {
-  name: string;
-  instance: ChildProcess | null;
-  stdout: any | null;
+export interface JavaEmulatorOptions {
   cacheDir: string;
   remoteUrl: string;
   expectedSize: number;
   expectedChecksum: string;
-  localPath: string;
   namePrefix: string;
 }
 
+export abstract class JavaEmulatorDetails {
```

Comment:
Previously this was just an `interface` but now that we have to support both `zip` and `jar` I found it easier to abstract this into a class where the emulators can specify multiple paths (not just the path where the download should go)

Reply:
Can we hide that as an implementation detail of `download`? These two classes look like they share a lot of logic.

Reply:
Ok you're right this sucked.  I changed it back to an interface but now I separated download constants from things like `instance` and `stdout` since they never really belonged together:
https://github.com/firebase/firebase-tools/pull/1755/commits/c58cdc6669a71e0b0b84b673cc9b5453895b84c8